test/grep.js

```diff
@@ -174,3 +174,166 @@ test('the pattern looks like an option', t => {
   t.falsy(shell.error());
   t.is(result.toString(), '-v\n-vv\n');
 });
+
+//
+// Before & after contexts
+//
+test('-B option', t => {
+  const result = shell.grep('-B', 3, 'test*', 'test/resources/grep/file3');
+  t.falsy(shell.error());
+  t.is(
+    result.toString(),
+    'line1\n' +
+      'line2 test line\n' +
+      'line3 test line\n' +
+      '--\n' +
+      'line7\n' +
+      'line8\n' +
+      'line9\n' +
+      'line10 test line\n' +
+      '--\n' +
+      'line12\n' +
+      'line13\n' +
+      'line14\n' +
+      'line15 test line\n'
+  );
+});
+
+test('-B option, -n option', t => {
+  const result = shell.grep('-nB', 3, 'test*', 'test/resources/grep/file3');
+  t.falsy(shell.error());
+  t.is(
+    result.toString(),
+    '1-line1\n' +
+      '2:line2 test line\n' +
+      '3:line3 test line\n' +
+      '--\n' +
+      '7-line7\n' +
+      '8-line8\n' +
+      '9-line9\n' +
+      '10:line10 test line\n' +
+      '--\n' +
+      '12-line12\n' +
+      '13-line13\n' +
+      '14-line14\n' +
+      '15:line15 test line\n'
+  );
+});
+
+test('-A option', t => {
+  const result = shell.grep('-A', 2, 'test*', 'test/resources/grep/file3');
+  t.falsy(shell.error());
+  t.is(
+    result.toString(),
+    'line2 test line\n' +
+      'line3 test line\n' +
+      'line4\n' +
+      'line5\n' +
+      '--\n' +
+      'line10 test line\n' +
+      'line11\n' +
+      'line12\n' +
+      '--\n' +
+      'line15 test line\n'
+  );
+});
+
+test('-A option, -B option', t => {
+  const result = shell.grep(
+    { '-A': 2, '-B': 3 },
+    'test*',
+    'test/resources/grep/file3'
+  );
+  t.falsy(shell.error());
+  t.is(
+    result.toString(),
+    'line1\n' +
+      'line2 test line\n' +
+      'line3 test line\n' +
+      'line4\n' +
+      'line5\n' +
+      '--\n' +
+      'line7\n' +
+      'line8\n' +
+      'line9\n' +
+      'line10 test line\n' +
+      'line11\n' +
+      'line12\n' +
+      'line13\n' +
+      'line14\n' +
+      'line15 test line\n'
+  );
+});
+
+test('-A option, -B option, -n option', t => {
+  const result = shell.grep(
+    { '-n': true, '-A': 2, '-B': 3 },
+    'test*',
+    'test/resources/grep/file3'
+  );
+  t.falsy(shell.error());
+  t.is(
+    result.toString(),
+    '1-line1\n' +
+      '2:line2 test line\n' +
+      '3:line3 test line\n' +
+      '4-line4\n' +
+      '5-line5\n' +
+      '--\n' +
+      '7-line7\n' +
+      '8-line8\n' +
+      '9-line9\n' +
+      '10:line10 test line\n' +
+      '11-line11\n' +
+      '12-line12\n' +
+      '13-line13\n' +
+      '14-line14\n' +
+      '15:line15 test line\n'
+  );
+});
+
+test('-C option', t => {
+  const result = shell.grep('-C', 3, 'test*', 'test/resources/grep/file3');
```

Comment:
Can you add a test case for `{'-C': 2}` or `{'-C': 1}`? I noticed these test cases use C values which are so large that the output doesn't include any line separators. That's actually a good thing to test, but we should additionally have a test for smaller C values just to verify the code does add the line separators when needed.

Reply:
I added a test case for `{'-C', 1}`.